atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializing.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.async.initializer;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.processors.DoDelegate;
+
+public interface AsyncInitializing {
+    @DoDelegate
+    default boolean isInitialized() {
+        return isInitializedAsync().isDone() && !isInitializedAsync().isCancelled();
+    }
+
+    @DoDelegate
+    default ListenableFuture<?> isInitializedAsync() {
```

Comment:
For this one, I _think_ I understand what you're trying to do, but we should discuss to make sure, and I think we might want to name it slightly different - maybe `getInitializationFuture` or similar. `isInitializedAsync` to me at least suggests "compute isInitialized in the background, and give me the result when it's done" which I don't think is what you intend.

Assuming the above is right, I can see two ways we can take this:
- if we don't want to expose the known-initialized version of the object to listeners once computation is done, we can probably just return a `ListenableFuture<Void>` unless I'm missing something
- if we do want to, should this return `T` and have the interface take a type parameter `T` as well?

Reply:
You make a good point. I was thinking about just "give me something to wait on until the object is initialized. Your second API would kind of make sense to me if we didn't expose the objects already, which we do all over the place.

As in the design of async init in Atlas is that you always get access to the underlying resource, BUT it throws exceptions until it's initialized.

To me ```ListenableFuture<?>``` actually carries this intent: it's just a notification thing. Lemme play around with actually returning ```<T>``` maybe that's actually what we want here.

Reply:
Alright so returning T is a much bigger effort, I think. The current decomp is such that Wrappers extend AsyncInitializer and they kind of just use it as a state tracker/state machine that they drive. So nowhere is T available.

However, I guess ```tryInitialize``` could return T, lemme try that.

Reply:
So #tryInitialize returning T, means that AsyncInitializable needs to have a generic parameter, so then everything that implements it needs to provide that. Which starts spiriling, and ultimately I don't feel like that's the right decomp.

We should RETURN ```AsyncInitializable<KeyValueService>``` from factories, not have KeyValueService extend that, I think this is my pet annoyance with the decomp here.

Lemme see how destructive this would be to just do for KeyValueService to prototype this cleanup (I'll have AtlasDbFactory force returning a wrapper, not the type itself). This is much closer to how Witchcraft returns Client interfaces from ConjureFactory and I think that's the right model here.

Reply:
Ultimately it feels like async init in atlas is just a false dichotomy as it's implemented: user's can choose to async initialize or not. But instead we should async initialize by default (because presumably it's faster since you can init unrelated pieces concurrently, instead of serially), but users should have a choice of whether they want to throw and block until it's done or let the rest of their system init.

Reply:
And I see we tried to do graphs of components, but it feels like in a rather clunky way, tieing initialization to the actual objects which pollutes the interfaces.

Reply:
Alright, having said all that what I'm proposing is obviously very complicated and totally beside the scope of this PR. I will rename the method to #getInitializationFuture() likely and call it a day and save this rambling for another day.